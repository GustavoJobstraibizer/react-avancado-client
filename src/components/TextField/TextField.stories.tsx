import { Meta, Story } from '@storybook/react'
import { Email } from '@styled-icons/material-outlined'
import TextField, { TextFieldProps } from '.'

export default {
  title: 'Form/TextField',
  component: TextField,
  args: {
    label: 'Name',
    labelFor: 'name',
    id: 'name',
    placeholder: 'Enter your name...'
  },
  argTypes: {
    onInput: { action: 'changed' },
    icon: { type: '' }
  }
} as Meta

export const Basic: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: '30rem', padding: '20px' }}>
    <TextField {...args} />
  </div>
)

export const withIcon: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: '30rem', padding: '20px' }}>
    <TextField {...args} />
  </div>
)

withIcon.args = {
  icon: <Email />
}

export const withError: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: '30rem', padding: '20px' }}>
    <TextField {...args} />
  </div>
)

withError.args = {
  error: 'Something went wrong...',
  icon: <Email />
}
