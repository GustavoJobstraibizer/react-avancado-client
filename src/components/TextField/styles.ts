import styled, { css, DefaultTheme } from 'styled-components'
import { TextFieldProps } from '.'

type IconPositionProps = Pick<TextFieldProps, 'iconPosition'>
type WrapperProps = Pick<TextFieldProps, 'disabled'> & { error?: boolean }

export const Label = styled.label`
  ${({ theme }) => css`
    cursor: pointer;
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
  `}
`
export const Input = styled.input<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    background: transparent;
    font-family: ${theme.font.family}
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
    padding-${iconPosition}: ${theme.spacings.xsmall};
    width: 100%;
    border: 0;
    outline: none;
    color: ${theme.colors.black};
  `}
`

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid ${theme.colors.lightGray};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`

export const Icon = styled.div<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    width: 2.2rem;
    display: flex;
    color: ${theme.colors.gray};

    order: ${iconPosition === 'right' ? 1 : 0};

    & > svg {
      width: 100%;
    }
  `}
`

const wrapperModifier = {
  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${Icon},
    ${Input} {
      cursor: not-allowed;
      color: ${theme.colors.gray};

      &::placeholder {
        color: currentColor;
      }
    }
  `,
  error: (theme: DefaultTheme) => css`
    ${InputWrapper} {
      border-color: ${theme.colors.red};
    }

    ${Label}, ${Icon} {
      color: ${theme.colors.red};
    }
  `
}

export const Wrapper = styled.main<WrapperProps>`
  ${({ theme, disabled, error }) => css`
    display: flex;
    flex-direction: column;

    ${disabled && wrapperModifier.disabled(theme)};
    ${error && wrapperModifier.error(theme)};
  `}
`

export const Error = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.red};
    font-size: 1.2rem;
  `}
`
