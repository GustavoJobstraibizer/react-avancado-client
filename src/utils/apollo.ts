import {
  ApolloClient,
  HttpLink,
  InMemoryCache,
  NormalizedCacheObject
} from '@apollo/client'
import { useMemo } from 'react'

const isSSR = typeof window === 'undefined'

let apolloClient: ApolloClient<NormalizedCacheObject>

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR,
    link: new HttpLink({ uri: 'http://localhost:1337/graphql' }),
    cache: new InMemoryCache()
  })
}

export function initializeApollo(initialState = {}) {
  // Verifica se ja existe uma instancia, se nao existir
  // cria uma nova se existir, usa a instancia existente
  const apolloClientGlobal = apolloClient ?? createApolloClient()

  // recupera os dados do cache
  if (initialState) {
    apolloClientGlobal.cache.restore(initialState)
  }

  // Inicializa no SSR com o cache limpo
  if (isSSR) return apolloClientGlobal

  apolloClient = apolloClient ?? apolloClientGlobal

  return apolloClient
}

export function useApollo(initialState = {}) {
  const store = useMemo(() => initializeApollo(initialState), [initialState])
  return store
}
