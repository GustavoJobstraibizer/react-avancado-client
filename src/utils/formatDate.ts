export const formatDate = (date: string): string => {
  return new Intl.DateTimeFormat('en-US', {
    day: 'numeric',
    month: 'short',
    year: 'numeric'
  }).format(new Date(date))
}
