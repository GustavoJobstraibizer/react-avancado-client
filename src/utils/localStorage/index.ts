const APP_KEY = 'WONGAMES'

const isServerSide = typeof window === 'undefined'

export function getStorageItem(key: string) {
  if (isServerSide) return

  const data = window.localStorage.getItem(`${APP_KEY}_${key}`)
  return data ? JSON.parse(data) : null
}

export function setStorageItem(key: string, values: string[]) {
  if (isServerSide) return

  window.localStorage.setItem(`${APP_KEY}_${key}`, JSON.stringify(values))
}
