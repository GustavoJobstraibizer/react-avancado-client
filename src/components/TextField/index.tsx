import React, { InputHTMLAttributes, useState } from 'react'
import * as S from './styles'

export type TextFieldProps = {
  label?: string
  initialValue?: string
  onInputChange?: (value?: string) => void
  icon?: React.ReactNode
  iconPosition?: 'left' | 'right'
  disabled?: boolean
  error?: string
} & InputHTMLAttributes<HTMLInputElement>

const TextField = ({
  label,
  initialValue = '',
  icon,
  onInputChange,
  name,
  iconPosition = 'left',
  disabled = false,
  error = '',
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue)

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.target.value

    setValue(newValue)

    if (onInputChange) {
      onInputChange(newValue)
    }
  }

  return (
    <S.Wrapper disabled={disabled} error={!!error}>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      <S.InputWrapper>
        {!!icon && <S.Icon iconPosition={iconPosition}>{icon}</S.Icon>}
        <S.Input
          iconPosition={iconPosition}
          type="text"
          onChange={onChange}
          value={value}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </S.InputWrapper>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  )
}

export default TextField
