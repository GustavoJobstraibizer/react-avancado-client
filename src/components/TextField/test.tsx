import { EmailOutline } from '@styled-icons/evaicons-outline/EmailOutline'
import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import theme from 'styles/theme'
import { renderWithTheme } from 'utils/test-utils'
import TextField from '.'

describe('TextField', () => {
  it('should render with label', () => {
    renderWithTheme(<TextField label="Name" labelFor="name" id="name" />)

    expect(screen.getByLabelText(/name/i)).toBeInTheDocument()
  })

  it('should render without label', () => {
    renderWithTheme(<TextField />)

    expect(screen.queryByLabelText(/name/i)).not.toBeInTheDocument()
  })

  it('should render with placeholder', () => {
    renderWithTheme(<TextField placeholder="Enter your name here" />)

    expect(
      screen.getByPlaceholderText(/Enter your name here/i)
    ).toBeInTheDocument()
  })

  it('should render with initial value', () => {
    renderWithTheme(<TextField initialValue="Hello" />)

    expect(screen.getByRole('textbox')).toHaveValue('Hello')
  })

  it('should change the value when user typing', async () => {
    const onInput = jest.fn()
    const valueTyping = 'Jhon doe'

    renderWithTheme(
      <TextField label="Name" labelFor="name" id="name" onInput={onInput} />
    )

    const textBox = screen.getByRole('textbox')

    expect(textBox).toBeInTheDocument()

    userEvent.type(textBox, valueTyping)

    await waitFor(() => {
      expect(textBox).toHaveValue(valueTyping)
      expect(onInput).toHaveBeenCalledTimes(valueTyping.length)
    })

    expect(onInput).toHaveBeenCalledWith(valueTyping)
  })

  it('should accessible by tab', () => {
    renderWithTheme(<TextField label="Name" labelFor="name" id="name" />)

    const textBox = screen.getByLabelText('Name')

    expect(document.body).toHaveFocus()

    userEvent.tab()

    expect(textBox).toHaveFocus()
    expect(textBox.parentElement).toHaveStyleRule(
      'box-shadow',
      `0 0 0.5rem ${theme.colors.primary}`,
      { modifier: ':focus-within' }
    )
  })

  it('should render with icon', () => {
    renderWithTheme(<TextField icon={<EmailOutline data-testid="icon" />} />)

    expect(screen.getByTestId('icon')).toBeInTheDocument()

    expect(screen.getByTestId('icon').parentElement).toHaveStyle({
      order: 0
    })
  })

  it('should render with icon to the right', () => {
    renderWithTheme(
      <TextField
        icon={<EmailOutline data-testid="icon" />}
        iconPosition="right"
      />
    )

    expect(screen.getByTestId('icon')).toBeInTheDocument()
    expect(screen.getByTestId('icon').parentElement).toHaveStyle({
      order: 1
    })
  })

  it('should render with disabled input and not allowing to typing', async () => {
    const onInput = jest.fn()

    renderWithTheme(<TextField disabled />)

    const text = 'test'
    const textBox = screen.getByRole('textbox')

    expect(textBox).toBeDisabled()
    expect(textBox).toHaveStyle({ cursor: 'not-allowed' })

    userEvent.type(textBox, text)

    await waitFor(() => {
      expect(textBox).not.toHaveValue(text)
    })

    expect(onInput).not.toHaveBeenCalled()
  })

  it('should not accessible by tab when disabled', () => {
    renderWithTheme(<TextField disabled />)

    const textBox = screen.getByRole('textbox')

    expect(textBox).toBeDisabled()

    expect(document.body).toHaveFocus()

    userEvent.tab()

    expect(textBox).not.toHaveFocus()
  })

  it('should render with error message', () => {
    const { container } = renderWithTheme(
      <TextField error="Something went wrong" label="E-mail" labelFor="email" />
    )

    expect(screen.getByText(/something went wrong/i)).toBeInTheDocument()
    expect(screen.getByRole('textbox').parentElement).toHaveStyle({
      'border-color': '#FF6347'
    })
    expect(screen.getByText(/e-mail/i)).toHaveStyle({
      color: '#FF6347'
    })

    expect(container.firstChild).toMatchSnapshot()
  })
})
