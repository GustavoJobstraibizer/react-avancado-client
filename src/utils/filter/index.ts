import { ItemProps } from 'components/ExploreSidebar'
import { ParsedUrlQueryInput } from 'querystring'

type ParseArgs = {
  queryString: ParsedUrlQueryInput
  filterItems?: Pick<ItemProps, 'type' | 'name'>[]
}

export const parseQueryStringToWhere = ({
  queryString,
  filterItems
}: ParseArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {}

  Object.keys(queryString)
    .filter((item) => item !== 'sort')
    .forEach((key) => {
      const item = filterItems?.find((i) => i.name === key)

      obj[key] = !isCheckbox(item)
        ? queryString[key]
        : { name_contains: queryString[key] }
    })

  return obj
}

export const parseQueryStringToFilter = ({
  queryString,
  filterItems
}: ParseArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {}

  Object.keys(queryString).forEach((key) => {
    const item = filterItems?.find((i) => i.name === key)
    const isArray = Array.isArray(queryString[key])

    obj[key] =
      !isArray && isCheckbox(item) ? [queryString[key]] : queryString[key]
  })

  return obj
}

const isCheckbox = (
  item: Pick<ItemProps, 'type' | 'name'> | undefined
): boolean => {
  return item?.type === 'checkbox'
}
