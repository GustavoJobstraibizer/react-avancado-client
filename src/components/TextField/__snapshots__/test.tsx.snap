// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField should render with error message 1`] = `
.c1 {
  cursor: pointer;
  font-size: 1.4rem;
  color: #030517;
}

.c4 {
  background: transparent;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif font-size:1.6rem;
  padding: 0.8rem 0;
  padding-left: 1.6rem;
  width: 100%;
  border: 0;
  outline: none;
  color: #030517;
}

.c4:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #EAEAEA inset;
  -webkit-filter: none;
  filter: none;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid #EAEAEA;
}

.c3:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .c2 {
  border-color: #FF6347;
}

.c5 {
  color: #FF6347;
  font-size: 1.2rem;
}

<main
  class="c0"
>
  <label
    class="c1"
    for="email"
  >
    E-mail
  </label>
  <div
    class="c2 c3"
  >
    <input
      class="c4"
      id="email"
      name="email"
      type="text"
      value=""
    />
  </div>
  <span
    class="c5"
  >
    Something went wrong
  </span>
</main>
`
